torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -765,68 +781,58 @@ def _get_ignored_modules(
-                    not isinstance(module, FlattenParamsWrapper), \
-                    "Ignoring FSDP modules is meaningless since their " \
-                    "parameters are not flattened into this FSDP module anyway"
-                for param_name, param in named_tensor_fn(module):
-                    prefixed_param_name = clean_param_name(prefix + param_name)
-                    ignored_named_tensors.add((prefixed_param_name, param))
-
-        def return_fn(ignored_named_tensors, *args):
-            return ignored_named_tensors
-
-        ignored_named_tensors = set()
-        return _apply_to_modules(
-            self, module_fn, return_fn, ignored_named_tensors, ignored_modules,
-        )
+    ) -> Set[torch.Tensor]:
+        """Returns the buffers of the modules in ``ignored_modules`` as a
+        :class:`set`."""
+        return set(b for m in ignored_modules for b in m.buffers())
 
-    def _get_ignored_named_parameters(self) -> Set[Tuple[str, torch.Tensor]]:
-        """Returns the named parameters of the modules in ``ignored_modules``,
-        excluding any :class:`FlatParameter` s."""
-        assert hasattr(self, "_ignored_modules"), \
-            "Expects `self._ignored_modules` to be initialized"
-        return self._get_ignored_named_tensors(
-            self._ignored_modules, lambda m: m.named_parameters(recurse=False),
-        )
-
-    def _get_ignored_named_buffers(self) -> Set[Tuple[str, torch.Tensor]]:
-        """Returns the named buffers of the modules in ``ignored_modules``,
-        excluding any :class:`FlatParameter` s."""
-        assert hasattr(self, "_ignored_modules"), \
-            "Expects `self._ignored_modules` to be initialized"
-        return self._get_ignored_named_tensors(
-            self._ignored_modules, lambda m: m.named_buffers(recurse=False),
+    def _get_ignored_tensor_names(
+        self,
+        module: torch.nn.Module,
+        ignored_params: Set[torch.nn.Parameter],
+        ignored_buffers: Set[torch.Tensor],
+    ) -> Set[str]:
+        """Returns a :class:`set` of ignored parameter and buffer names."""
+        param_to_unflat_param_names = _get_param_to_unflat_param_names(module)
+        ignored_param_names = set()
+        for param in ignored_params:
+            unflat_param_names = param_to_unflat_param_names[param]
+            assert len(unflat_param_names) == 1, \
+                "Only `FlatParameter`s can map to >1 unflattened parameter " \
+                "name, and `_get_ignored_parameters()` should have excluded " \
+                "them; check `_get_param_to_unflat_param_names()`"
+            ignored_param_names.add(unflat_param_names[0])
+        buffer_to_buffer_name = _get_buffer_to_buffer_name(module)
+        ignored_buffer_names = set(
+            buffer_to_buffer_name[b] for b in ignored_buffers
         )
+        assert ignored_param_names.isdisjoint(ignored_buffer_names), \
```

Comment:
Do we need this assert even though we end up taking the union afterwards? Agreed it is for correctness, but also wondering the performance tradeoff. If it is done only once in ctor as I think it is, it should be fine though.